ProcessLib/HydroMechanics/HydroMechanicsFEM.h

```diff
@@ -0,0 +1,425 @@
+    // compilers.
+    explicit IntegrationPointData(IntegrationPointData&& other)
+        : _b_matrices(std::move(other._b_matrices)),
+          _sigma(std::move(other._sigma)),
+          _sigma_prev(std::move(other._sigma_prev)),
+          _eps(std::move(other._eps)),
+          _eps_prev(std::move(other._eps_prev)),
+          _solid_material(other._solid_material),
+          _material_state_variables(std::move(other._material_state_variables)),
+          _C(std::move(other._C)),
+          integration_weight(std::move(other.integration_weight))
+    {
+    }
+#endif  // _MSC_VER
+
+    typename ShapeMatrixTypeDisplacement::NodalRowVectorType _N_u;
+    typename BMatricesType::BMatrixType _b_matrices;
+    typename BMatricesType::KelvinVectorType _sigma, _sigma_prev;
+    typename BMatricesType::KelvinVectorType _eps, _eps_prev;
+
+    typename ShapeMatricesTypePressure::NodalRowVectorType _N_p;
+    typename ShapeMatricesTypePressure::GlobalDimNodalMatrixType _dNdx_p;
+
+    MaterialLib::Solids::MechanicsBase<DisplacementDim>& _solid_material;
+    std::unique_ptr<typename MaterialLib::Solids::MechanicsBase<
+        DisplacementDim>::MaterialStateVariables>
+        _material_state_variables;
+
+    typename BMatricesType::KelvinMatrixType _C;
+    double integration_weight;
+
+    void pushBackState()
+    {
+        _eps_prev = _eps;
+        _sigma_prev = _sigma;
+        _material_state_variables->pushBackState();
+    }
+
+    template <typename DisplacementVectorType>
+    void updateConstitutiveRelation(double const t,
+                                    SpatialPosition const& x_position,
+                                    double const dt,
+                                    DisplacementVectorType const& u)
+    {
+        _eps.noalias() = _b_matrices * u;
+        _solid_material.computeConstitutiveRelation(
+            t, x_position, dt, _eps_prev, _eps, _sigma_prev, _sigma, _C,
+            *_material_state_variables);
+    }
+};
+
+struct HydroMechanicsLocalAssemblerInterface
+    : public ProcessLib::LocalAssemblerInterface
+{
+};
+
+template <typename ShapeFunctionDisplacement, typename IntegrationMethod,
+          int DisplacementDim>
+class HydroMechanicsLocalAssembler
+    : public HydroMechanicsLocalAssemblerInterface
+{
+public:
+    using ShapeMatricesType =
+        ShapeMatrixPolicyType<ShapeFunctionDisplacement, DisplacementDim>;
+
+    // Types for displacement.
+    // (Higher order elements = ShapeFunctionDisplacement).
+    using NodalMatrixType = typename ShapeMatricesType::NodalMatrixType;
+    using BMatricesType =
+        BMatrixPolicyType<ShapeFunctionDisplacement, DisplacementDim>;
+
+    using BMatrixType = typename BMatricesType::BMatrixType;
+
+    // Types for pressure.
+    // (Lower order elements = Order(ShapeFunctionDisplacement) - 1).
+    using ShapeFunctionPressure =
+        typename LowerDim<ShapeFunctionDisplacement>::type;
```

Comment:
I would add `ShapeFunctionDisplacement` to template parameters of this class and use LowerDim in LocalDataInitializer, because algorithms in the local assembler don't need this information.